### ISSUE TITLE
DBT rpc hangs when child process is killed.

### ISSUE THREAD

**Issue by davidbraden** [fishtown-analytics/dbt#3114]
Current behaviour
When running a dbt rpc server,  if the child process spawned to run the task is killed by an OOM error it hangs until the timeout (if set). In particular if no timeout is set it will hang indefinitely. 

Expected behaviour
The dbt rpc server detects if the spawned task has died and errors immediately. 

https://github.com/fishtown-analytics/dbt/blob/develop/core/dbt/rpc/logger.py#L165

**Comment by jtcohen6** [comment]
@davidbraden Could you give a little more context to help understand what's going wrong here:
- How are you deploying the RPC server?
- What's causing the OOM error?
- How do you determine the child task has been killed by an OOM error? By polling the child task, examining logs from the deployment mechanism, ...?

> Expected behaviour
> The dbt rpc server detects if the spawned task has died and errors immediately.

To your mind, does this look like server polling its running child tasks on a regular interval, and raising an error or killing as soon as the poll is unsuccessful? Today, this is something the server expects the client to do, via `ps` + `poll` + `kill`.

**Comment by davidbraden** [comment]
The RPC server is deployed as a docker container running on k8s in google cloud. 

The OOM error was caused because the container had only been given 2GB of memory and at some point in the processing the container exceeded that limit. 

The OOM was detected because we saw "Memory cgroup out of memory: Kill process " in the k8s cluster logs. At the same point in the dbt logs the job started hanging with no more debug output. The incident was investigated at a later time when the container had been restarted so there was no option of running `ps` and seeing the processes. 

k8s has a container healthcheck but it only checks the root process of the container which container to return healthy. See https://github.com/kubernetes/kubernetes/issues/78973 
I think it might make sense in the loop https://github.com/fishtown-analytics/dbt/blob/develop/core/dbt/rpc/logger.py#L165 to instead poll for a smaller interval than the full timeout and then check if the process is running. If the process has stopped then it should break the loop.

**Comment by jtcohen6** [comment]
@davidbraden I appreciate the context, that makes a lot of sense. If I understand right, the proposed change here is that, in addition to `timeout`, each RPC method should accept a common parameter like `poll_interval`, with a default setting of `none` or a reasonable interval (30 seconds?). If `poll_interval` is set, then `dispatch_until_exit` would check the child task at each interval (and return an exception if there is one), rather than once only when `timeout` is reached.

Is this a code change you' be interested in contributing?

**Comment by davidbraden** [comment]
Yeah that's the change I'm proposing. Unfortunately I probably won't get round to contributing to it anytime soon. Hopefully this context will be still useful if anyone has similar issues.
